ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -515,21 +515,31 @@ SelectValueMapper getSelectValueMapper() {
 
   @SuppressWarnings("unchecked")
   public SchemaKStream<?> selectKey(
-      final Field newKeyField,
+      final KeyField selectKeyField,
       final boolean updateRowKey,
       final QueryContext.Stacker contextStacker
   ) {
-    final Optional<Field> resolved = keyField.resolve(schema, ksqlConfig);
+    final Field field = selectKeyField.resolve(schema, ksqlConfig)
+        .orElseThrow(IllegalArgumentException::new);
 
-    final boolean namesMatch = resolved
-        .map(kf -> SchemaUtil.matchFieldName(kf, newKeyField.name()))
+    final KeyField newKeyField =
```

Comment:
it took me a surprising amount of time to grok this code (especially in relation to the rest of the method). I think what tripped me up is the `keyField.withLegacy(selectKeyField.legacy())`. This splits the key, but the split is only really needed for the `if (namesMatch ...)` clause (as evidenced by the basically duplicated condition), but is only used if the namesMatch part of the condition matches. 

Instead, I think we should expose this cyclomatic complexity in explicit if/else (or if/return if you so desire) branches instead of hiding it under conditional state:
```
if (names match) {
  // use selectKeyField
} else if (usingNewKeys && isRowKey()) {
  // use the existing keyField (doesn't matter what legacy is set to)
} else {
  // do the repartition
}
```
Feel free to do something else instead of my suggestion, but I do request exposing the structure in control blocks instead of state. This complexity is also compounded by the fact that at this point it is very possible that the legacy key is null (empty) but the new key will never be empty.

Reply:
I'll have a play and see what works.  I agree: code that's easy to read is good code.

Reply:
So....

>  I think what tripped me up is the keyField.withLegacy(selectKeyField.legacy()). This splits the key, but the split is only really needed for the if (namesMatch ...) clause 

This is not the cause, the result is used for all paths.

I've reworked it very slightly and renamed some vars to hopefully make things clearer.